cirq/experiments/qubit_characterizations.py

```diff
@@ -0,0 +1,544 @@
+
+    Returns:
+        A StateTomographyResult object that stores and plots the density matrix.
+    """
+
+    def _measurement(two_qubit_circuit: circuits.Circuit) -> Sequence[float]:
+        two_qubit_circuit.append(ops.measure(q_0, q_1, key='z'))
+        results = sampler.run(two_qubit_circuit, repetitions=num_shots)
+        bit_strings = [r for r in results.measurements['z']]
+        p_00 = 0.0
+        p_01 = 0.0
+        p_10 = 0.0
+        for bits in bit_strings:
+            if not bits[0] and not bits[1]:
+                p_00 += 1.0 / num_shots
+            elif not bits[0] and bits[1]:
+                p_01 += 1.0 / num_shots
+            elif bits[0] and not bits[1]:
+                p_10 += 1.0 / num_shots
+        return [p_00, p_01, p_10]
+
+    sigma_0 = numpy.eye(2) / 2.0
+    sigma_1 = numpy.array([[0.0, 1.0], [1.0, 0.0]]) / 2.0
+    sigma_2 = numpy.array([[0.0, -1.0j], [1.0j, 0.0]]) / 2.0
+    sigma_3 = numpy.array([[1.0, 0.0], [0.0, -1.0]]) / 2.0
+    sigmas = [sigma_0, sigma_1, sigma_2, sigma_3]
+
+    probs = []
+    rots = [ops.X ** 0, ops.X ** 0.5, ops.Y ** 0.5]
+    mat = numpy.zeros((27, 15))
+    s = numpy.array([[1.0, 1.0, 1.0], [-1.0, 1.0, -1.0], [1.0, -1.0, -1.0]])
+
+    for i, rot_1 in enumerate(rots):
+        for j, rot_2 in enumerate(rots):
+            m_idx, indices, signs = _indices_after_basis_rot(i, j)
+            mat[m_idx: (m_idx + 3), indices] = s * numpy.tile(signs, (3, 1))
+            test_circuit = circuit + circuits.Circuit.from_ops(rot_1(q_1))
+            test_circuit.append(rot_2(q_0))
+            probs.extend(_measurement(test_circuit))
+
+    c, _, _, _ = numpy.linalg.lstsq(mat, 4.0 * numpy.array(probs) - 1.0,
+                                    rcond=None)
+    c = numpy.concatenate(([1.0], c))
+    c = c.reshape(4, 4)
+
+    rho = numpy.zeros((4, 4))
+    for i in range(4):
+        for j in range(4):
+            rho = rho + c[i, j] * numpy.kron(sigmas[i], sigmas[j])
+
+    return TomographyResult(rho)
+
+
+def _indices_after_basis_rot(i, j):
+    mat_idx = 3 * (3 * i + j)
+    q_0_i = 3 - i
+    q_1_j = 3 - j
+    indices = [q_1_j - 1, 4 * q_0_i - 1, 4 * q_0_i + q_1_j - 1]
+    signs = [(-1) ** (j == 2), (-1) ** (i == 2), (-1) ** (i == 2 + j == 2)]
+    return mat_idx, indices, signs
+
+
+def _two_qubit_clifford_matrices(q_0, q_1, c1, s1, s1_x, s1_y):
+    mats = []
+    for i in range(11520):
+        circuit = circuits.Circuit()
+        circuit.append(_two_qubit_clifford(q_0, q_1, i, c1, s1, s1_x, s1_y))
+        mats.append(protocols.unitary(circuit))
+    return numpy.array(mats)
+
+
+def _random_single_q_clifford(qubit, num_cfds, cfds, cfd_matrices):
+    gate_ids = list(numpy.random.choice(24, num_cfds))
+    gate_sequence = []
+    for gate_id in gate_ids:
+        gate_sequence.extend(cfds[gate_id])
+    idx = _find_inv_matrix(_gate_seq_to_mats(gate_sequence), cfd_matrices)
+    gate_sequence.extend(cfds[idx])
+    circuit = circuits.Circuit()
```

Comment:
`Circuit.from_ops(gate(qubit) for gate in gate_sequence)`

Reply:
Done